### ISSUE TITLE
SUCI ProfileB error, iPhone can't connect

### ISSUE THREAD

**Issue by jao-tm** [magma/magma#15444]
### Discussed in https://github.com/magma/magma/discussions/15320

<div type='discussions-op-text'>

<sup>Originally posted by **jao-tm** October 11, 2023</sup>
Hi all,
Has anyone tested the SUCI features? I've tried to implement it, but get the error message `subscriberdb[2145211]: ERROR:grpc._server:Exception calling application: An X25519 private key is 32 bytes long` when I try to attach. Key pairs has been generated with `openssl genpkey -algorithm X25519` and added to UE sim card and Magma network. In the documentation I can only find the [SUCI Extensions](https://magma.github.io/magma/docs/next/lte/suci_extensions) page which doesn't provide any debug steps.</div>

**Comment by Radula76** [comment]
We have only ever used Profile A with iPhones.

**Comment by kmu-tm** [comment]
Hi. 

I'd like to update this ticket with new information. We have conducted various tests for both Profile A and Profile B using an iPhone (testing each profile separately and together when provisioned on the SIM card). These Profile A and Profile B tests were performed using the 3GPP values mentioned in TS 133 501. We have verified the length of the keys, the protection scheme, and other parameters on both the NMS and the SIM card.

[Logs](https://pastebin.com/raw/r7dYJ839) captured when we tested with IPhone using Profile A

Our initial thought was maybe the key that are added in the NMS front-end has a length that does not match with what is added to the back-end (?). But this fix was already patched in a prior release, as mentioned in this [ticket](https://github.com/magma/magma/issues/13708). 

[Logs ](https://pastebin.com/raw/aDsmSkKY)captured when we tested with IPhone using Profile B

From my understanding for Profile B, it seems like the deserialization of keys is not working properly B. I will update this ticket with more logs and experiment other scenarios tomorrow. In the meantime, please do let us know on how we can proceed with this test.

**Comment by kmu-tm** [comment]
**Small update:** We have verified the key provisioned in the NMS is the same as what is stored in the SubscriberDB. These public and private key is from 3GPP test data (for Profile B).  This implies that the keys remain unchanged. Please see the following:

home_net_public_key_id: 2
protection_scheme     : 1
home_net_public_key   : 0272da71976234ce833a6907425867b82e074d44ef907dfb4b3e21c1c2256ebcd1
home_net_private_key  : f1ab1074477ebcc7f554ea1c5fc368b1616730155e0041ac447d6301975fecda

I believe there might be some issues with the cryptography code for decrypting the keys that are stored in the database. We will continue to debug further.

**Comment by brunohcfaria** [comment]
The problem is in the format of the HN private key that is not DEC/PCKS8. I've made a patch to fix that and soon I'll open a PR.

**Comment by kmu-tm** [comment]
> The problem is in the format of the HN private key that is not DEC/PCKS8. I've made a patch to fix that and soon I'll open a PR.

Yes,  we investigated this SUCI key format issue earlier today. Our findings was that for Profile B (secp256r1), the cryptographic library currently accepts the private key in DER format, while the public key works in raw hex format. Under these conditions, the deserialization function operates correctly, although this has only been verified with test scripts, not with actual UEs (I can provide a new update on Monday using IPhone). For Profile A, the expected format is raw 32-byte for both private and public keys, with no requirement for DER formatting. However, issues persist with Profile A that require further investigation.

It is important to note that the current setup for Profile B deviates from the SUCI documentation (from Magma), which specifies the use of raw format for keys. To ensure alignment with the documentation, we should consider updating our implementation to support raw hex format for both profiles, or even update the documentation (?).

Any insights on the Profile A issue would be greatly appreciated. For Profile A, the problem is still with `An X25519 private key is 32 bytes long` (using 3GPP values). I need to debug more on the NMS/AGW side.

**Comment by brunohcfaria** [comment]
> 
> Yes, we investigated this SUCI key format issue earlier today. Our findings was that for Profile B (secp256r1), the cryptographic library currently accepts the private key in DER format, while the public key works in raw hex format. Under these conditions, the deserialization function operates correctly, although this has only been verified with test scripts, not with actual UEs (I can provide a new update on Monday using IPhone). For Profile A, the expected format is raw 32-byte for both private and public keys, with no requirement for DER formatting. However, issues persist with Profile A that require further investigation.

Take a look [here](https://github.com/brunohcfaria/magma/tree/fix_profile_b_suci_decrypt) for a fix. Can you try it and give feedback on the ProfileB key length error?

> It is important to note that the current setup for Profile B deviates from the SUCI documentation (from Magma), which specifies the use of raw hex format for both keys. To ensure alignment with the documentation, we should consider updating the EC implementation to support raw hex format for both profiles, or perhaps update the documentation (?).
> 
> Any insights on the Profile A issue would be greatly appreciated. For Profile A, the problem is still with `An X25519 private key is 32 bytes long` (using 3GPP values). I need to debug more on the NMS/AGW side.

Ok, I've looked only into ProfileB issue (probably because I've kept the issue title in my mind). 
For ProfileA I'll have to check the behavior.

**Comment by kmu-tm** [comment]
> Take a look [here](https://github.com/brunohcfaria/magma/tree/fix_profile_b_suci_decrypt) for a fix. Can you try it and give feedback on the ProfileB key length error?
> 
Hi @brunohcfaria. Thanks for your efforts. I think there might be some confusion here. The key length error is **not** for Profile B, but for Profile A. [Here](https://pastebin.com/raw/awvFu87h) are the updated logs for Profile B when using the correct format for key generation. Based on these logs, we can see that the MSIN decryption is not working properly. The correct IMSI (the UE that is attempting to attach) is 999700000068375.

I have tested the public and private key (for Profile B) using test scripts to ensure that the crypto library is working, and from my tests it seems like everything is OK. With that said, there is no problem with the key format.

**Comment by kmu-tm** [comment]
Hi. I thought I'd update this ticket with more information, perhaps other people might find this useful (so others do not go in a loophole when debugging with SUCI).

As I mentioned in my previous reply, I have tested using a test script that makes use of the gRPC setup and fetches the credentials from the database. Our findings was that using the MSIN and protection scheme, the script is able to decrypt the last 10 digits. This implies that there is **no** issue with the cryptography library, gRPC, database, or the private and public key (that are provisioned from the NMS), as shown below:

![MSIN decryption](https://github.com/user-attachments/assets/9130c644-f827-40f0-8858-7e44567b2fb2)

The problem is mostly on the AMF side with MSIN decoding. The logs show "000099970008000" instead of 999700000068375. We have tested the decoding patch by bruno, but now we are getting "get_decrypt_imsi_info fails with code 3, Message : Public key length or protection scheme is invalid
033834 Tue Aug 20 07:07:12 2024 7F43FF7FE700 ERROR AMF-AP lib/n11/M5GSUCIRegistrationServi:0068  Deconcealing IMSI Failed, sending Registration Reject". Please keep in mind that we are testing with a real handset and not a simulator.

**Comment by kmu-tm** [comment]
The issue is now fixed in the PR https://github.com/magma/magma/pull/15499. We can successfully confirm that SUCI using Profile B is working with IPhone and IPad.  This ticket can now be closed.

**Comment by lucasgonze** [comment]
Reopening because the state of the merge isn't clear.
